specs/sharding/polynomial-commitments.md

```diff
@@ -0,0 +1,387 @@
+- [Helper functions](#helper-functions)
+    - [`next_power_of_two`](#next_power_of_two)
+- [Field operations](#field-operations)
+  - [Generic field operations](#generic-field-operations)
+    - [`bls_modular_inverse`](#bls_modular_inverse)
+    - [`roots_of_unity`](#roots_of_unity)
+  - [Field helper functions](#field-helper-functions)
+    - [`compute_powers`](#compute_powers)
+    - [`low_degree_check`](#low_degree_check)
+    - [`vector_lincomb`](#vector_lincomb)
+    - [`bytes_to_field_elements`](#bytes_to_field_elements)
+- [Polynomial operations](#polynomial-operations)
+    - [`add_polynomials`](#add_polynomials)
+    - [`multiply_polynomials`](#multiply_polynomials)
+    - [`interpolate_polynomial`](#interpolate_polynomial)
+    - [`evaluate_polynomial_in_evaluation_form`](#evaluate_polynomial_in_evaluation_form)
+- [KZG Operations](#kzg-operations)
+  - [Elliptic curve helper functoins](#elliptic-curve-helper-functoins)
+    - [`elliptic_curve_lincomb`](#elliptic_curve_lincomb)
+  - [Hash to field](#hash-to-field)
+    - [`hash_to_bls_field`](#hash_to_bls_field)
+  - [KZG operations](#kzg-operations)
+    - [`verify_kzg_proof`](#verify_kzg_proof)
+    - [`verify_kzg_multiproof`](#verify_kzg_multiproof)
+    - [`verify_degree_proof`](#verify_degree_proof)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+<!-- /TOC -->
+
+
+## Introduction
+
+This document specifies basic polynomial operations and KZG polynomial commitment operations as they are needed for the sharding specification. The implementations are not optimized for performance, but readability. All practical implementations shoul optimize the polynomial operations, and hints what the best known algorithms for these implementations are are included below.
+
+## Constants
+
+### BLS Field
+
+| Name | Value | Notes |
+| - | - | - |
+| `BLS_MODULUS` | `0x73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001` (curve order of BLS12_381) |
+| `PRIMITIVE_ROOT_OF_UNITY` | `7` | Primitive root of unity of the BLS12_381 (inner) BLS_MODULUS |
+
+### KZG Trusted setup
+
+| Name | Value |
+| - | - |
+| `G1_SETUP` | Type `List[G1]`. The G1-side trusted setup `[G, G*s, G*s**2....]`; note that the first point is the generator. |
+| `G2_SETUP` | Type `List[G2]`. The G2-side trusted setup `[G, G*s, G*s**2....]` |
+
+## Custom types
+
+We define the following Python custom types for type hinting and readability:
+
+| Name | SSZ equivalent | Description |
+| - | - | - |
+| `KZGCommitment` | `Bytes48` | A G1 curve point |
+| `BLSFieldElement` | `uint256` | A number `x` in the range `0 <= x < BLS_MODULUS` |
+| `BLSPolynomialByCoefficients` | `List[BLSFieldElement]` | A polynomial over the BLS field, given in coefficient form |
+| `BLSPolynomialByEvaluations` | `List[BLSFieldElement]` | A polynomial over the BLS field, given in evaluation form |
+
+## Helper functions
+
+#### `next_power_of_two`
+
+```python
+def next_power_of_two(x: int) -> int:
```

Comment:
This seems to bug out if called with `x=0`:

```
In [24]: next_power_of_two(0)
Out[24]: 2

In [25]: next_power_of_two(1)
Out[25]: 1

In [26]: next_power_of_two(2)
Out[26]: 2
```

Perhaps it will never be called with x=0 but might be worth noting down or asserting for it.